### ISSUE TITLE
Name resolution through proxy server?

### ISSUE THREAD

**Issue by ovidiubuligan** [Karasiq/proxychain#2]
Hello,
I am trying to use this behind a proxy. And I get:

```
java.net.UnknownHostException: This is usually a temporary error during hostname resolution and means that the local ser
ver did not receive a response from an authoritative server
        at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
        at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:907)
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1302)
        at java.net.InetAddress.getAllByName0(InetAddress.java:1255)
        at java.net.InetAddress.getAllByName(InetAddress.java:1171)
        at java.net.InetAddress.getAllByName(InetAddress.java:1105)
        at java.net.InetAddress.getByName(InetAddress.java:1055)
        at com.karasiq.proxychain.FirewallImpl.connectionIsAllowed(Firewall.scala:49)
        at com.karasiq.proxychain.app.Handler$$anonfun$com$karasiq$proxychain$app$Handler$$waitConnection$1.applyOrElse(
Handler.scala:210)
```

Does name resolution also go through the proxy ? 

Wha I want to do is to create a chain of 2 proxyes : 
 `proxy1 -> proxy2` All name resolution has to go through the chain.

If I want to debug this issue and try a quick fix , would you have any suggestion where in the code to look at?

**EDIT**:
This my fist time running any scala code.
I set up these 2 propperties right after `object Boot extends App {`

```
  System.setProperty("sun.net.spi.nameservice.nameservers", "10.0.1.1");
  System.setProperty("sun.net.spi.nameservice.provider.1", "dns,sun");
```

and now I get a different error 

```
[ERROR] [05/26/2016 13:17:28.794] [ProxyChain-akka.actor.default-dispatcher-4] [akka://ProxyChain/user/proxychain-server/$a] DNS error
java.net.UnknownHostException: DNS error
    at sun.net.spi.nameservice.dns.DNSNameService.resolve(DNSNameService.java:180)
    at sun.net.spi.nameservice.dns.DNSNameService.lookupAllHostAddr(DNSNameService.java:331)
```

It treats it as a dns server which is probably not, it's just a squid proxy.

**Comment by Karasiq** [comment]
Local DNS resolve is used for filtering and can be disabled in config https://github.com/Karasiq/proxychain/blob/master/src/main/resources/reference.conf#L21

**Comment by ovidiubuligan** [comment]
Hello,
Setting **allowDNS** to false gives me this `scala.MatchError: www.google.com:443 (of class java.net.InetSocketAddress)` Scala Match error : 

```
[ERROR] [05/30/2016 09:44:25.844] [ProxyChain-akka.actor.default-dispatcher-9] [akka://ProxyChain/user/proxychain-server/$c] www.google.com:443 (of class java.net.InetSocketAddress)
scala.MatchError: www.google.com:443 (of class java.net.InetSocketAddress)
    at com.karasiq.proxychain.FirewallImpl.connectionIsAllowed(Firewall.scala:55)
    at com.karasiq.proxychain.app.Handler$$anonfun$com$karasiq$proxychain$app$Handler$$waitConnection$1.applyOrElse(Handler.scala:210)
    at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:170)
    at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:467)
    at com.karasiq.proxychain.app.Handler.aroundReceive(Handler.scala:32)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by Karasiq** [comment]
Fixed in 4d1628bf691d99daea4716c04d6ed19d1cfe4f27

**Comment by ovidiubuligan** [comment]
Hello, 
Running with the latest code  i get: (in eclipse debug)

```
[INFO] [06/01/2016 08:54:11.679] [ProxyChain-akka.actor.default-dispatcher-3] [Proxy(akka://ProxyChain)] Opened connection through proxy chain [http://user:pa$$@proxy.company.com:3128 -> http://myproxyoninternet:8081] to google.ro:80
```

but nothing shows up in the browser.Don't know hot to enable higher logging.
If I change the dollar sign in the password of the first proxy string I get:

```
HTTP/1.0 408 Request Time-out
Cache-Control: no-cache
Connection: close
Content-Type: text/html

<html><body><h1>408 Request Time-out</h1>
Your browser didn't send a complete request in time.
</body></html>
```

For now I created a working docker image for proxychaining  : https://github.com/ovidiubuligan/tinyproxychain.
For me any solution is good but  I am also interested in memory usage and performance of this scala solution vs tinyproxy + proxychains in a Vbox. Do you think scala solution is more performant?

**Comment by Karasiq** [comment]
Are you using the 2.0.3 version? All tests works as expected.
![screenshot 2016-06-01 001](https://cloud.githubusercontent.com/assets/7199719/15704429/66487042-27f3-11e6-992b-45fbcdcd878d.png)

`[INFO] [06/01/2016 12:11:07.907] [ProxyChain-akka.actor.default-dispatcher-21] [Proxy(akka://ProxyChain)] Opened connection through proxy chain [http://User_232352386:123456@127.0.0.1:9999 -> http://109.124.95.77:8080 -> socks://95.46.11.142:1080 -> socks://my-exit-proxy] to 2ip.ru:443`
![screenshot 2016-06-01 002](https://cloud.githubusercontent.com/assets/7199719/15704400/39fa20d0-27f3-11e6-9cc2-f0075f41d789.png)

**Comment by ovidiubuligan** [comment]
Well it seams that with the latest code it works ,,,, But.
Testing with firefox works only with existing tabs that were oppened with docker proxy solution(also youtube works).  But if I open a new tab and type youtube.com it  doesn't connect.

I probably missconfigured the ssl certs ?
What I did is generate a  jks and copy and paste it and assigned it to both `trust-store` and `key-store`.

If this is the case .. can you add some steps to set up the keys in readme.md ? Are the ssl certs mandatory ?
**EDIT:**
With the firefox testing case, it might still use the old proxy setting (docker)
accessing a `http` site with the latest code gives me:
This page:

```
HTTP/1.0 408 Request Time-out
Cache-Control: no-cache
Connection: close
Content-Type: text/html

<html><body><h1>408 Request Time-out</h1>
Your browser didn't send a complete request in time.
</body></html>
```

And this error in console:

```
[ERROR] [06/01/2016 16:57:37.885] [ProxyChain-akka.actor.default-dispatcher-15] [akka://ProxyChain/user/StreamSupervisor-0/flow-28-0-unknown-operation] Error in stage [com.karasiq.proxy.client.HttpProxyClientStage@736f2edb]: requirement failed: Cannot push port (HttpProxyClient.dataOut) twice
java.lang.IllegalArgumentException: requirement failed: Cannot push port (HttpProxyClient.dataOut) twice
    at scala.Predef$.require(Predef.scala:224)
    at akka.stream.stage.GraphStageLogic.push(GraphStage.scala:435)
    at com.karasiq.proxy.client.HttpProxyClientStage$$anon$1$$anon$2.onPush(HttpProxyClientStage.scala:48)
    at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:587)
    at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:598)
    at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:539)
    at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:472)
    at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:424)
    at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:604)
    at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:619)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:482)
    at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:530)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
    at akka.actor.ActorCell.invoke(ActorCell.scala:495)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
    at akka.dispatch.Mailbox.run(Mailbox.scala:224)
    at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by Karasiq** [comment]
> I probably missconfigured the ssl certs ?

Most likely you don't need the SSL settings, it's used for TLS-proxy server, which is not supported by browsers.

> requirement failed: Cannot push port (HttpProxyClient.dataOut) twice

Ok, try the fixed version 75ef30c6ee6a37cc6d6184777868ddf8857f0866

> HTTP/1.0 408 Request Time-out

Is it response from your HTTP proxy, or from google? I doubt that google uses HTTP/1.0. Have you tried the other proxy servers?
![screenshot 2016-06-01 004](https://cloud.githubusercontent.com/assets/7199719/15715238/5ffd4856-2826-11e6-979c-8ed726c152d7.png)

**Comment by ovidiubuligan** [comment]
Hello,(disclamer this project is my first experience with scala ).
I tested with the latest version ad now I don't get any logs to the console on connection. Putting breakpoints on different places hits themwhile running and oppening the browser. 

`HTTP/1.0 408 Request Time-out` heppend on `olx.ro` site , on google the page remains blank.
![image](https://cloud.githubusercontent.com/assets/4757459/15736008/8e492520-28a5-11e6-9f54-4aa080e38410.png)

My proxy chain is : 
Authenticated squid proxy with http 1.1 (i did a curl -v on it)  ->  Privoxy proxy with http1.1 (I think https://www.privoxy.org/)  

The thing is that for the authenticated proxy , in the password I have this `$` character , escaping it with `%24` gives me the same result. Running nmap on the authenticated proxy shows me that ipv6 is not supported. **I needed to add  `System.setProperty("java.net.preferIPv4Stack" , "true");` in Boot.scala to work**  
First proxy from the `entry` config is : `http://user:pa%24@proxy.company.com:3128` 
Second proxy from the `exit` config is : `http://myIp:8081`.

Can you upload on gist or send me a confing I can use in which I can plug in just the `entry` and `exit` urls?

**Comment by Karasiq** [comment]
> I tested with the latest version ad now I don't get any logs to the console on connection

Try the version with logging 3742489c29fcc72a9e4f382182df3c2dd2807d71
It should output something like

```
[DEBUG] [06/02/2016 12:21:56.795] [proxy-chain-test-akka.actor.default-dispatcher-2] [ProxyStage(akka://proxy-chain-test)] Sending request to HTTP proxy http://127.0.0.1:9999: CONNECT ipecho.net:80 HTTP/1.1
Host: ipecho.net


[DEBUG] [06/02/2016 12:21:56.801] [proxy-chain-test-akka.actor.default-dispatcher-4] [akka://proxy-chain-test/system/IO-TCP/selectors/$a/0] Attempting connection to [/127.0.0.1:9999]
[DEBUG] [06/02/2016 12:21:56.803] [proxy-chain-test-akka.actor.default-dispatcher-4] [akka://proxy-chain-test/system/IO-TCP/selectors/$a/0] Connection established to [/127.0.0.1:9999]
[DEBUG] [06/02/2016 12:21:56.951] [proxy-chain-test-akka.actor.default-dispatcher-6] [ProxyStage(akka://proxy-chain-test)] Received response from HTTP proxy http://127.0.0.1:9999: HTTP/1.1 200 Connection established (HC)


[DEBUG] [06/02/2016 12:21:56.951] [proxy-chain-test-akka.actor.default-dispatcher-6] [ProxyStage(akka://proxy-chain-test)] Connection established to ipecho.net:80 through HTTP proxy: http://127.0.0.1:9999
```

**Comment by ovidiubuligan** [comment]
I placed both proxies in the `entry` config and commented aout the middle and exit  because it showed List(List(p1,p2),List(p1,p2))  but even with one both proxies In one `entry` config it gives me : 

```
[DEBUG] [06/02/2016 13:06:13.878] [ProxyChain-akka.actor.default-dispatcher-7] [Proxy(akka://ProxyChain)] Trying connect to ipecho.net:80 through chains: List(List(http://my_authed_company_proxy:3128, http://my_internet_proxy:8081), List(http://my_authed_company_proxy:3128, http://my_internet_proxy:8081))
[DEBUG] [06/02/2016 13:06:13.900] [ProxyChain-akka.actor.default-dispatcher-8] [akka://ProxyChain/system/IO-TCP/selectors/$a/12] Attempting connection to [my_authed_company_proxy/10.0.1.11:3128]
[DEBUG] [06/02/2016 13:06:13.901] [ProxyChain-akka.actor.default-dispatcher-2] [akka://ProxyChain/system/IO-TCP/selectors/$a/13] Attempting connection to [my_authed_company_proxy/10.0.1.11:3128]
[DEBUG] [06/02/2016 13:06:13.956] [ProxyChain-akka.actor.default-dispatcher-2] [akka://ProxyChain/system/IO-TCP/selectors/$a/12] Connection established to [my_authed_company_proxy/10.0.1.11:3128]
[DEBUG] [06/02/2016 13:06:13.961] [ProxyChain-akka.actor.default-dispatcher-6] [akka://ProxyChain/system/IO-TCP/selectors/$a/13] Connection established to [my_authed_company_proxy/10.0.1.11:3128]
[INFO] [06/02/2016 13:06:14.345] [ProxyChain-akka.actor.default-dispatcher-8] [Proxy(akka://ProxyChain)] Opened connection through proxy chain [http://my_authed_company_proxy:3128 -> http://my_internet_proxy:8081] to ipecho.net:80
```

After you change to the timeout I now get a timeout in the browser after 30 seconds.

**Comment by Karasiq** [comment]
Are you sure that the correct version of `proxyutils` library is used? Try running with `sbt clean run`
It must log the HTTP response, before "Opened connection through proxy chain" line: https://github.com/Karasiq/proxyutils/blob/master/src/main/scala/com/karasiq/proxy/client/HttpProxyClientStage.scala#L48-L55

---

Also i ran a test with Squid -> SOCKS5 -> Privoxy -> TOR chain, and it works as expected

```
[DEBUG] [06/02/2016 16:19:59.233] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Requesting available auth methods from SOCKS proxy: socks://127.0.0.1:1080
[DEBUG] [06/02/2016 16:19:59.234] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Sending request to HTTP proxy http://127.0.0.1:8118: CONNECT 127.0.0.1:9050 HTTP/1.1
Host: /127.0.0.1:9050


[DEBUG] [06/02/2016 16:19:59.234] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Sending request to HTTP proxy http://u$er:pa$$@127.0.0.1:3128: CONNECT 127.0.0.1:1080 HTTP/1.1
Proxy-Authorization: Basic dSRlcjpwYSQk
Host: /127.0.0.1:1080


[DEBUG] [06/02/2016 16:19:59.234] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Sending request to HTTP proxy http://127.0.0.1:9050: CONNECT ipecho.net:80 HTTP/1.1
Host: ipecho.net


[DEBUG] [06/02/2016 16:19:59.235] [proxy-chain-test-akka.actor.default-dispatcher-5] [akka://proxy-chain-test/system/IO-TCP/selectors/$a/2] Attempting connection to [/127.0.0.1:3128]
[DEBUG] [06/02/2016 16:19:59.236] [proxy-chain-test-akka.actor.default-dispatcher-5] [akka://proxy-chain-test/system/IO-TCP/selectors/$a/2] Connection established to [/127.0.0.1:3128]
[DEBUG] [06/02/2016 16:19:59.239] [proxy-chain-test-akka.actor.default-dispatcher-2] [ProxyStage(akka://proxy-chain-test)] Received response from HTTP proxy http://u$er:pa$$@127.0.0.1:3128: HTTP/1.1 200 Connection established


[DEBUG] [06/02/2016 16:19:59.239] [proxy-chain-test-akka.actor.default-dispatcher-2] [ProxyStage(akka://proxy-chain-test)] Connection established to /127.0.0.1:1080 through HTTP proxy: http://u$er:pa$$@127.0.0.1:3128
[DEBUG] [06/02/2016 16:19:59.240] [proxy-chain-test-akka.actor.default-dispatcher-4] [ProxyStage(akka://proxy-chain-test)] SOCKS proxy has no authentication: socks://127.0.0.1:1080
[DEBUG] [06/02/2016 16:19:59.240] [proxy-chain-test-akka.actor.default-dispatcher-4] [ProxyStage(akka://proxy-chain-test)] Sending connection request to SOCKS proxy socks://127.0.0.1:1080 -> /127.0.0.1:8118
[DEBUG] [06/02/2016 16:19:59.243] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Connection established to /127.0.0.1:8118 through SOCKS proxy: socks://127.0.0.1:1080
[DEBUG] [06/02/2016 16:19:59.248] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Received response from HTTP proxy http://127.0.0.1:8118: HTTP/1.1 200 Connection established


[DEBUG] [06/02/2016 16:19:59.249] [proxy-chain-test-akka.actor.default-dispatcher-3] [ProxyStage(akka://proxy-chain-test)] Connection established to /127.0.0.1:9050 through HTTP proxy: http://127.0.0.1:8118
[DEBUG] [06/02/2016 16:19:59.254] [proxy-chain-test-akka.actor.default-dispatcher-2] [ProxyStage(akka://proxy-chain-test)] Received response from HTTP proxy http://127.0.0.1:9050: HTTP/1.0 200 Connection established
Content-Length: 0


[DEBUG] [06/02/2016 16:19:59.254] [proxy-chain-test-akka.actor.default-dispatcher-2] [ProxyStage(akka://proxy-chain-test)] Connection established to ipecho.net:80 through HTTP proxy: http://127.0.0.1:9050
HTTP/1.1 200 OK
Date: Thu, 02 Jun 2016 13:19:55 GMT
Server: Apache
Expires: Mon, 26 Jul 1997 05:00:00 GMT
Cache-Control: no-cache
Pragma: no-cache
Vary: Accept-Encoding
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Transfer-Encoding: chunked
Content-Type: text/html

d
171.25.193.77
0
```

```
1464873599.523    286 127.0.0.1 TCP_TUNNEL/200 419 CONNECT 127.0.0.1:1080 u$er HIER_DIRECT/127.0.0.1 -
```

```
2016-06-02 16:19:59.243 00002b0c Request: 127.0.0.1:9050/
```

**Comment by ovidiubuligan** [comment]
I tested with a clean sbt clean run and got this https://gist.github.com/ovidiubuligan/ab2abbf50988729d734428016c4bbf0f
The thing is that Authorization header should be sent only for the first proxy (the 3128 port one) and I think it is also sent to the second one. The second proxy doesn't have any authentification.
The squid proxi (3128 port one ) is "corporate configured" .. but tinyproxy+ proxychains (from inside docker vm)  with default configs can go trough itit seams .

**Comment by Karasiq** [comment]
First proxy request

```
CONNECT my_proxy:8081 HTTP/1.1
Proxy-Authorization: Basic AAAAAAAAAAAAAAAAAA
Host: /my_proxy:8081
```

```
HTTP/1.0 200 Connection established
```

Second proxy request

```
CONNECT clients6.google.com:443 HTTP/1.1
Host: clients6.google.com:443
```

```
HTTP/1.1 200 Connection established
```

Seems like both proxies are connected, but request hangs. Try to append one of these proxies https://gist.github.com/Karasiq/f0e6eaada93e96e31b0ccea0043d35e3 to the chain, or replace the :8081 proxy and connect again.

**Comment by ovidiubuligan** [comment]
I tested wih one of the provided public servers and I get the same:
https://gist.github.com/ovidiubuligan/7e745db80d2ab76e36e6ca1dcaf2d162.
I want to try to emulate tinyproxy headders (dump them with wireshark from the vm interface) .What do I need to to add more headders sent to the first proxy?.

**Comment by Karasiq** [comment]
The request is generated here
https://github.com/Karasiq/proxyutils/blob/master/src/main/scala/com/karasiq/proxy/client/HttpProxyClientStage.scala#L34-L39
Also try the recent fixes: db609f729dd99822e3cd8c291446285deb471b6d
